porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go

```diff
@@ -777,3 +782,158 @@ func getFileKubeObject(prr *porchapi.PackageRevisionResources, file, kind, name
+		existingmutators = fn.SliceSubObjects{}
+	}
+
+	for _, mutator := range existingmutators {
+		ok, err := isPackageVariantFunc(mutator, pv.ObjectMeta.Name)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			newMutators = append(newMutators, mutator)
+		}
+	}
+
+	var newPVMutators = fn.SliceSubObjects{}
+	for i, mutator := range pv.Spec.Pipeline.Mutators {
+		newMutator := mutator.DeepCopy()
+		newMutator.Name = generatePVFuncName(mutator.Name, pv.ObjectMeta.Name, i)
+		mut, err := fn.NewFromTypedObject(newMutator)
+		if err != nil {
+			return err
+		}
+		newPVMutators = append(newPVMutators, &mut.SubObject)
+	}
+
+	newMutators = append(newPVMutators, newMutators...)
+
+	// generate new validators
+	var newValidators = fn.SliceSubObjects{}
+
+	existingvalidators, ok, err := pipeline.NestedSlice("validators")
+	if err != nil {
+		return err
+	}
+	if !ok || existingvalidators == nil {
+		existingvalidators = fn.SliceSubObjects{}
+	}
+
+	for _, validator := range existingvalidators {
+		ok, err := isPackageVariantFunc(validator, pv.ObjectMeta.Name)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			newValidators = append(newValidators, validator)
+		}
+	}
+
+	var newPVValidators = fn.SliceSubObjects{}
+	for i, validator := range pv.Spec.Pipeline.Validators {
+		newValidator := validator.DeepCopy()
+		newValidator.Name = generatePVFuncName(validator.Name, pv.ObjectMeta.Name, i)
+		val, err := fn.NewFromTypedObject(newValidator)
+		if err != nil {
+			return err
+		}
+		newPVValidators = append(newPVValidators, &val.SubObject)
+	}
+
+	newValidators = append(newPVValidators, newValidators...)
+
+	// update kptfile
+	if len(newMutators) > 0 {
+		if err := pipeline.SetSlice(newMutators, "mutators"); err != nil {
+			return err
+		}
+	}
+
+	if len(newValidators) > 0 {
+		if err := pipeline.SetSlice(newValidators, "validators"); err != nil {
+			return err
+		}
+	}
+
+	prr.Spec.Resources[kptfilev1.KptFileName] = kptfile.String()
+
+	return nil
+}
+
+const PackageVariantFuncPrefix = "PackageVariant"
+
+// isPackageVariantFunc returns true if a function has been created via a PackageVariant.
+// It uses the name of the func to determine its origin and compares it with the supplied pvName.
+func isPackageVariantFunc(fn *fn.SubObject, pvName string) (bool, error) {
+	origname, ok, err := fn.NestedString("name")
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve field name: %w", err)
+	}
+	if !ok {
+		return false, fmt.Errorf("could not find field name in supplied func")
+	}
+
+	name := strings.Split(origname, ".")
+
+	// if more or less than 3 dots have been used, return false
+	if len(name) != 4 {
+		return false, nil
+	}
+
+	// if PackageVariantFuncPrefix has not been used, return false
+	if name[0] != PackageVariantFuncPrefix {
+		return false, nil
+	}
+
+	// if pv-names don't match, return false
+	if name[1] != pvName {
+		return false, nil
+	}
+
+	// make sure func name is not empty
+	if name[2] == "" {
```

Comment:
I think it can be empty? The user doesn't need to add a name in the PV function name field, so this would seem to be a problem if they didn't. See the "(if present)" here: https://github.com/GoogleContainerTools/kpt/blob/main/docs/design-docs/08-package-variant.md#kptfile-function-pipeline-editing

Or am I missing something, perhaps there is a need for this or you put a default value in ?

Reply:
I see. Yes should be fine. 

Just to clarify the generated name would then look like this right?: `Packagevariant.my-pv..0`

Reply:
Either that or `PackageVariant.my-pv.0` works for me. Just make sure the design is consistent with the implementation.